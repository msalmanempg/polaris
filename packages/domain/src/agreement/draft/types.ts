export interface DraftAgreement {
  id?: number;
  agreementNumber: string;
  unitId: number;
  data: any;
  createdAt?: Date;
  updatedAt?: Date;
}
