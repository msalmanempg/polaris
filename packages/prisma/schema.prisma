//******** AUTO GENERATED FILE, DO NOT EDIT.  *********

datasource db {
    provider = "postgresql"
    url      = env("DATABASE_URL")
}

generator client {
    provider = "prisma-client-js"
}

generator dbml {
    provider = "prisma-dbml-generator"
}


model Agreement {
  id                   Int                 @id @default(autoincrement())
  leadId               Int                 @map("lead_id")
  agreementNumber      String              @unique @map("agreement_number")
  publishedPrice       Int                 @map("published_price")
  bookingPrice         Int                 @map("booking_price")
  discountValue        Int?                @map("discount_value")
  posessionValue       Int                 @map("posession_value")
  payments             ExpectedPayment[]
  customers            AgreementCustomer[]
  salesPerson          SalePerson[]
  nominees             Nominee[]
  Unit                 Unit                @relation(fields: [unitId], references: [id])
  unitId               Int                 @map("unit_id")
  status               agreementStatus
  bookingDate          DateTime            @map("booking_date")
  notarisedDate        DateTime            @map("notarised_date")
  agreementDate        DateTime            @map("agreement_date")
  installmentStartDate DateTime            @map("installment_start_date")
  installmentEndDate   DateTime            @map("installment_end_date")
  createdAt            DateTime            @default(now()) @map("created_at")
  updatedAt            DateTime            @updatedAt @map("updated_at")

  @@map("agreements")
}

model DraftAgreement {
  id              Int      @id @default(autoincrement())
  Unit            Unit     @relation(fields: [unitId], references: [id])
  unitId          Int      @map("unit_id")
  agreementNumber String   @unique @map("agreement_number")
  data            Json     @map("data")
  createdAt       DateTime @default(now()) @map("created_at")
  updatedAt       DateTime @updatedAt @map("updated_at")

  @@map("draft_agreements")
}

model SalePerson {
  id          Int       @id @default(autoincrement())
  designation String
  fullName    String    @map("full_name")
  email       String
  govtId      String    @map("govt_id")
  phoneNumber String    @map("phone_number")
  sortLevel   Int       @map("sort_level")
  Agreement   Agreement @relation(fields: [agreementId], references: [id])
  agreementId Int       @map("agreement_id")
  createdAt   DateTime  @default(now()) @map("created_at")
  updatedAt   DateTime  @updatedAt @map("updated_at")

  @@map("sale_persons")
}

model AgreementCustomer {
  id                   Int                @id @default(autoincrement())
  type                 customerType       @map("customer_type")
  Customer             Customer           @relation(fields: [customerId], references: [id])
  customerId           Int                @map("customer_id")
  Agreement            Agreement          @relation(fields: [agreementId], references: [id])
  agreementId          Int                @map("agreement_id")
  isPrimary            Boolean            @default(true)
  meta                 Json?
  email                String
  primaryPhoneNumber   String             @map("primary_phone_number")
  secondaryPhoneNumber String?            @map("secondary_phone_number")
  address              String
  Companies            AgreementCompany[]
  createdAt            DateTime           @default(now()) @map("created_at")
  updatedAt            DateTime           @updatedAt @map("updated_at")

  @@map("agreement_customers")
}

model AgreementCompany {
  id                  Int               @id @default(autoincrement())
  AgreementCustomer   AgreementCustomer @relation(fields: [agreementCustomerId], references: [id])
  agreementCustomerId Int               @map("agreement_customer_id")
  name                String
  ntn                 String
  registrationNumber  String            @unique @map("registration_number")
  address             String
  city                String
  email               String
  contactNumber       String            @map("contact_number")
  meta                Json?
  createdAt           DateTime          @default(now()) @map("created_at")
  updatedAt           DateTime          @updatedAt @map("updated_at")

  @@map("agreement_companies")
}

model Nominee {
  id            Int        @id @default(autoincrement())
  govtId        String     @map("govt_id")
  govtIdType    govtIdType @map("govt_id_card")
  nomineeFor    String     @map("nominee_for")
  fullName      String     @map("full_name")
  contactNumber String     @map("contact_number")
  email         String
  relationship  String
  Agreement     Agreement  @relation(fields: [agreementId], references: [id])
  agreementId   Int        @map("agreement_id")
  meta          Json?
  createdAt     DateTime   @default(now()) @map("created_at")
  updatedAt     DateTime   @updatedAt @map("updated_at")

  @@map("nominees")
}

model AgreementAttachment {
  id           Int                   @id @default(autoincrement())
  agreementId  Int                   @map("agreement_id")
  Attachment   Attachment            @relation(fields: [attachmentId], references: [id])
  attachmentId Int
  type         agreementDocumentType
  createdAt    DateTime              @default(now()) @map("created_at")
  updatedAt    DateTime              @updatedAt @map("updated_at")

  @@map("agreement_attachments")
}

model NomineeAttachment {
  id           Int                   @id @default(autoincrement())
  nomineeId    Int                   @map("nominee_id")
  Attachment   Attachment            @relation(fields: [attachmentId], references: [id])
  attachmentId Int
  type         nomineeAttachmentType
  createdAt    DateTime              @default(now()) @map("created_at")
  updatedAt    DateTime              @updatedAt @map("updated_at")

  @@map("nominee_attachment")
}

enum govtIdType {
  cnic
  nicop
  poc
}

enum customerType {
  individual
  company
}

enum agreementStatus {
  draft
  review
  active
  transferred
  terminated
}

enum agreementDocumentType {
  agreement
  booking_form
  payment_plan
  floor_plan
  other
}

enum nomineeAttachmentType {
  id_front
  id_back
  profile_picture
  other
}


model Attachment {
  id                   Int                   @id @default(autoincrement())
  name                 String
  key                  String
  customerAttachments  CustomerAttachment[]
  nomineeAttachments   NomineeAttachment[]
  agreementAttachments AgreementAttachment[]
  createdAt            DateTime              @default(now()) @map("created_at")
  updatedAt            DateTime              @updatedAt @map("updated_at")

  @@map("attachments")
}


model Customer {
    id             Int                  @id @default(autoincrement())
    fullName       String               @map("full_name")
    govtId         String               @map("govt_id")
    govtIdType     govtIdType           @map("govt_id_type")
    relationType   relationType         @map("relation_type")
    guardianName   String               @map("guardian_name")
    nationality    String
    passportNumber String?              @map("passport_number")
    dateOfBirth    DateTime             @map("date_of_birth")
    gender         Gender
    meta           Json?
    attachments    CustomerAttachment[]
    agreements     AgreementCustomer[]
    createdAt      DateTime             @default(now()) @map("created_at")
    updatedAt      DateTime             @updatedAt @map("updated_at")

    @@unique([govtId, govtIdType], name: "Customer_govtId_govtIdType_unique_constraint")
    @@map(name: "customers")
}

model CustomerAttachment {
    id           Int        @id @default(autoincrement())
    Attachment   Attachment @relation(fields: [attachmentId], references: [id])
    attachmentId Int
    Customer     Customer   @relation(fields: [customerId], references: [id])
    customerId   Int
    createdAt    DateTime   @default(now()) @map("created_at")
    updatedAt    DateTime   @updatedAt @map("updated_at")

    @@map("customer_attachments")
}

enum relationType {
    father
    husband
}

enum Gender {
    male
    female
}

enum customerDocumentType {
    id_front
    id_back
    profile_picture
    other
}


model ExpectedPayment {
  id          Int                   @id @default(autoincrement())
  amount      Int
  percentage  Int?
  paymentDate DateTime              @map("payment_date")
  type        expectedPaymentType
  Agreement   Agreement             @relation(fields: [agreementId], references: [id])
  agreementId Int                   @map("agreement_id")
  meta        Json?
  dueDate     DateTime              @map("due_date")
  status      expectedPaymentStatus
  createdAt   DateTime              @default(now()) @map("created_at")
  updatedAt   DateTime              @updatedAt @map("updated_at")

  @@map("expected_payments")
}

enum expectedPaymentType {
  down_payment
  installment
  advance
  posession
}

enum expectedPaymentStatus {
  paid
  pending
}


model Project {
  id             Int                 @id @default(autoincrement())
  name           String
  address        String
  city           String
  province       String
  country        String
  longitude      Float
  latitude       Float
  email          String
  phone          String
  url            String
  completionDate DateTime?           @map("completion_date")
  logo           String
  meta           Json?
  bankDetails    ProjectBankDetail[]
  units          Unit[]
  createdAt      DateTime            @default(now()) @map("created_at")
  updatedAt      DateTime            @updatedAt @map("updated_at")

  @@map("projects")
}

model ProjectBankDetail {
  id            Int      @id @default(autoincrement())
  bankName      String   @map("bank_name")
  accountTitle  String   @map("account_title")
  accountNumber String   @map("account_number")
  iban          String
  swiftCode     String   @map("swift_code")
  branchAddress String   @map("branch_address")
  branchCode    String   @map("branch_code")
  meta          Json?
  Project       Project? @relation(fields: [projectId], references: [id])
  projectId     Int?
  createdAt     DateTime @default(now()) @map("created_at")
  updatedAt     DateTime @updatedAt @map("updated_at")

  @@map("project_bank_details")
}


model Unit {
  id             Int              @id @default(autoincrement())
  unitNumber     String           @unique @map("unit_number")
  location       String
  type           unitType
  bed            Int
  basePrice      Int              @map("base_price")
  publishedPrice Int              @map("published_price")
  netArea        Int              @map("net_area")
  grossArea      Int              @map("gross_area")
  meta           Json?
  status         unitStatus
  Project        Project          @relation(fields: [projectId], references: [id])
  projectId      Int              @map("project_id")
  agreements     Agreement[]
  drafts         DraftAgreement[]
  completionDate DateTime?        @map("completion_date")
  createdAt      DateTime         @default(now()) @map("created_at")
  updatedAt      DateTime         @updatedAt @map("updated_at")

  @@map(name: "units")
}

enum unitType {
  residential
  commercial
}

enum unitStatus {
  booked
  sold
  available
}
