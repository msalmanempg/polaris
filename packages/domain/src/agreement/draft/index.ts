export { DraftAgreementService } from "./draft-agreement-service";
export * from "./types";
