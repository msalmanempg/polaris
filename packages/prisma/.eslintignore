**/__snapshots__/
**/.ng_build/
**/build/
**/coverage/
**/dist/
**/lib/
**/node_modules/
jest.config.js
*.d.ts