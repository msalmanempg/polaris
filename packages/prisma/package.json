{
  "name": "@polaris/prisma",
  "version": "1.0.0",
  "description": "data layer for Polaris",
  "main": "lib/index.js",
  "repository": "https://github.com/SectorLabs/polaris/tree/main/packages/prisma",
  "license": "MIT",
  "scripts": {
    "build": "yarn ttsc -p tsconfig.json",
    "lint:errors": "eslint --quiet --ext .js,.ts ./",
    "lint:types": "ttsc --noEmit"
  }
}
