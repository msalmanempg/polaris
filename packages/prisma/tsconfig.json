{
  "compilerOptions": {
    "allowJs": true,
    "module": "commonjs",
    "esModuleInterop": true,
    "moduleResolution": "node",
    "skipLibCheck": true,
    "isolatedModules": true,
    "forceConsistentCasingInFileNames": true,
    "lib": ["es2017", "dom"],
    "target": "es2017",
    "types": ["node"],
    "strict": true,
    "baseUrl": ".",
    "paths": {
      "prisma/*": ["."]
    },
    "outDir": "./lib/",
    "sourceMap": true,
    "inlineSources": true,
    "sourceRoot": "/",
    "plugins": [{ "transform": "@zerollup/ts-transform-paths" }]
  },
}
