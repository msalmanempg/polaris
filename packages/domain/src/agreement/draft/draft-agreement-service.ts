import { PrismaClient } from "@polaris/prisma";
import { Injectable } from "injection-js";
import { DraftAgreement } from "./types";

@Injectable()
export class DraftAgreementService {
  constructor(private prisma: PrismaClient) {}

  /**
   * Create Draft Agreement
   * @param { DraftAgreement } data draft data
   * @returns { Promise<DraftAgreement> } Promise containing newly created draft
   */
  async create(data: DraftAgreement): Promise<DraftAgreement> {
    return this.prisma.draftAgreement.upsert({
      where: { agreementNumber: data.agreementNumber },
      create: { ...data },
      update: { ...data },
    });
  }

  /**
   * Get All Draft Agreements
   * @returns {  Promise <DraftAgreement[]> } Promise containing list of all drafts
   */
  async all(): Promise<DraftAgreement[]> {
    return this.prisma.draftAgreement.findMany({});
  }

  /**
   * Get Detail of a Draft by Id or Agreement Number
   * @param { number } id unique record id
   * @param { string } agreementNumber unique agreementNumber
   * @returns { Promise<DraftAgreement>} Promise containing record details
   */
  async getDetails(id?: number, agreementNumber?: string): Promise<DraftAgreement | null> {
    const query = id ? { id: id } : agreementNumber ? { agreementNumber: agreementNumber } : {};
    return this.prisma.draftAgreement.findUnique({ where: { ...query } });
  }

  /**
   * Remove record from table
   * @param { string } agreementNumber unique agreement number
   * @returns { Promise<DraftAgreement>} Promise containing deleted record details
   */
  async delete(agreementNumber: string): Promise<DraftAgreement | void> {
    return this.prisma.draftAgreement.delete({ where: { agreementNumber: agreementNumber } });
  }
}
