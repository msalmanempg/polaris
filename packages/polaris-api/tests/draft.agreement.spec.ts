import { prisma, Project, Unit } from "@polaris/prisma";
import { projects, units } from "./support/fixtures/units";
import { draftAgreements } from "./support/fixtures/draft.agreements";
import { testExpressApp } from "./support/requestFactory";

describe("draftAgreements", () => {
  let createdProject: Project;
  let createdUnits: Unit[];
  beforeEach(async () => {
    const [project] = projects;

    createdProject = await prisma.project.create({
      data: {
        ...(project as any),
      },
    });

    createdUnits = await Promise.all(
      units.map((unit: any) => {
        return prisma.unit.upsert({
          where: { unitNumber: unit.unitNumber },
          update: { ...unit, id: undefined, projectId: createdProject.id },
          create: { ...unit, id: undefined, projectId: createdProject.id, status: "available" },
        });
      })
    );

    await Promise.all(
      draftAgreements.map((draft) => {
        const data = {
          ...draft,
          Unit: { connect: { id: createdUnits[0].id } },
          unitId: undefined,
        } as any;
        return prisma.draftAgreement.create({ data: data });
      })
    );
  });

  afterEach(async () => {
    await prisma.draftAgreement.deleteMany({});
    await prisma.unit.deleteMany({
      where: {
        projectId: {
          in: createdProject.id,
        },
      },
    });
    await prisma.project.deleteMany({ where: { id: createdProject.id } });
  });

  it("Should create a new agreement draft", async () => {
    const rawDraft = { agreementNumber: "ABC1236", unitId: createdUnits[0].id, data: {} };

    const { body } = await testExpressApp().post("/api/agreement/draft").send(rawDraft).expect(201);

    expect(body).toMatchSnapshot({
      id: expect.any(Number),
      createdAt: expect.any(String),
      updatedAt: expect.any(String),
      unitId: expect.any(Number),
    });
  });

  it("Should get draft agreement details", async () => {
    const draft = await prisma.draftAgreement.findUnique({
      where: { agreementNumber: draftAgreements[0].agreementNumber },
    });

    const { body } = await testExpressApp()
      .get(`/api/agreement/draft/detail/${draft ? draft.id : 0}`)
      .expect(200);
    expect(body).toMatchSnapshot({
      id: expect.any(Number),
      createdAt: expect.any(String),
      updatedAt: expect.any(String),
      unitId: expect.any(Number),
    });
  });
});
