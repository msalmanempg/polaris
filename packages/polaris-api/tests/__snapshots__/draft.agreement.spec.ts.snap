// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`draftAgreements Should create a new agreement draft 1`] = `
Object {
  "agreementNumber": "ABC1236",
  "createdAt": Any<String>,
  "data": Object {},
  "id": Any<Number>,
  "unitId": Any<Number>,
  "updatedAt": Any<String>,
}
`;

exports[`draftAgreements Should get draft agreement details 1`] = `
Object {
  "agreementNumber": "ABC1234",
  "createdAt": Any<String>,
  "data": Object {},
  "id": Any<Number>,
  "unitId": Any<Number>,
  "updatedAt": Any<String>,
}
`;
